---
title: "An introduction to the atena package"
author:
- name: Beatriz Calvo-Serra
  affiliation:
  - &id Dept. of Experimental and Health Sciences, Universitat Pompeu Fabra, Barcelona, Spain
  email: beatriz.calvo@upf.edu
- name: Robert Castelo
  affiliation: *id
  email: robert.castelo@upf.edu
package: "`r pkg_ver('atena')`"
abstract: >
  The `atena` package provides methods to quantify the expression of transposable elements within R and Bioconductor.
vignette: >
  %\VignetteIndexEntry{An introduction to the atena package}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
output:
  BiocStyle::html_document:
    toc: true
    toc_float: true
    number_sections: true
bibliography: bibliography.bib
---

```{r setup, echo=FALSE}
options(width=80)
```

# What are transposable elements

Transposable elements (TEs) are autonomous mobile genetic elements. They are DNA sequences that have, or once had, the ability to mobilize within the genome, either directly or through an RNA intermediate [@payer2019transposable]. TEs can be categorized into two classes, based on the intermediate substrate propagating insertions (RNA or DNA). Class I TEs, also called retrotransposons, first transcribe an RNA copy that is then reverse transcribed to cDNA before inserting in the genome. In turn, these can be divided into long terminal repeat (LTR) retrotransposons, which refer to endogenous retroviruses (ERVs), and non-LTR retrotransposons, which include long interspersed element class 1 (LINE-1 or L1) and short interspersed elements (SINEs). Class II TEs, also known as DNA transposons, directly excise themselves from one location before reinsertion. TEs are further split into families and subfamilies depending on various structural features [@goerner2018computational]  [@guffanti2018novel].

Most TEs have lost the capacity for generating new insertions over their evolutionary history and are now fixed in the human population. Their insertions have resulted in a complex distribution of interspersed repeats comprising almost half (50%) of the human genome [@payer2019transposable].

TE expression has been observed in association with physiological processes in a wide range of species, including humans where it has been described to be important in early embryonic pluripotency and development. Moreover, aberrant TE expression has been associated with diseases such as cancer, neurodegenerative disorders, and infertility [@payer2019transposable].

# Current available methods for quantifying TE expression

The study of TE expression faces one main challenge: given their repetitive nature, the majority of TE-derived reads map to multiple regions of the genome and these multi-mapping reads are consequently discarded in standard RNA-seq data processing pipelines. For this reason, specific software for the quantification of TE expression have been developed [@goerner2018computational].

Available methods for quantifying TE expression can be classified into family-based or locus-based. The former combines expression at the subfamily level (e.g. *TEtranscripts*), whereas the latter provides TE expression at the locus level (e.g. *ERVmap* and *Telescope*). In both cases, there are different approaches to estimate TE expression:

* Tools based on selective filtering of multi-mapping reads, such as [ERVmap](https://github.com/mtokuyama/ERVmap). **ERVmap** applies multiple filtering criteria for RNA-sequencing reads. These filters consist in discarding reads when the ratio of sum of hard and soft clipping to the length of the read (base pair) is greater than or equal to 0.02, the ratio of the edit distance to the sequence read length (base pair) is greater or equal to 0.02 and/or the difference between the alignment score from BWA (field AS) and the suboptimal alignment score from BWA (field XS) is less than 5 [@tokuyama2018ervmap]. 

* Tools based on reassignment of ambiguously mapped reads, such as [Telescope](https://github.com/mlbendall/telescope) and [TEtranscripts](https://github.com/mhammell-laboratory/TEtranscripts). **Telescope** uses a Bayesian model with an expectation-maximization method to estimate the probability of read assignment and reassign multi-mapping reads to the most probable source transcript [@bendall2019telescope]. **TEtranscripts** also uses an expectation-maximization (EM) algorithm to determine the maximum-likelihood estimates of multi-reads assignments to TEs but, finally, addresses the sequence similarities between TEs from the same subfamily by combining TE expression at the subfamily level [@jin2015tetranscripts].

Because these tools were only available outside R and Bioconductor, the `atena` package provides access to these three methods to facilitate the integration of TE expression quantification into Bioconductor workflows for the analysis of RNA-seq data.


# TEs annotations

Another challenge in TE expression quantification is the lack of complete TE annotations due to the difficulty to correctly place TEs in genome assemblies [@goerner2018computational]. The gold standard for TE annotations in human are RepeatMasker annotations, available through the RepeatMasker tracks in genome browsers. Other repositories offer annotations for other species (see *Table 1* in [@goerner2018computational]).

In this package, we include annotations for 3,220 human ERVs (HERVs) (a type of class I TEs) collected by ERVmap authors [@tokuyama2018ervmap], available [here](https://github.com/mtokuyama/ERVmap/blob/master/ERVmap.bed) (GRCh38/hg38 genome assembly). Moreover, this package also contains annotations of TEs (HERVs and L1 long interspersed nuclear element (LINE)) created by Telescope authors [@bendall2019telescope] for the GRCh38/hg38 human genome assembly (28,513 TEs) and for the GRCh37/hg19 assembly (28,314 TEs), available [here](https://github.com/mlbendall/telescope_annotation_db).


# Using atena to quantify TE expression

Quantification of TE expression with `atena` consists in the following two steps:

1. Building of a parameter object for one of the available quantification methods.

2. Calling of the TE expression quantification method `qtex()` using the previously
   built parameter object.

## Building a parameter object for ERVmap

As mentioned, `atena` implements the ERVmap method for TE expression quantification. In order to use this approach, first, an object of the class `ERVmapParam` is build using the `ERVmapParam()` function. In this example, the annotations of 3,220 HERVs included in this package are used. The `singleEnd` argument is set to `TRUE` since the example BAM files are single-end. The `ignoreStrand` argument is used to define if the read counting method (`summarizeOverlaps` from the [*GenomicAlignments*](https://bioconductor.org/packages/release/bioc/html/GenomicAlignments.html) package) takes the strand of reads and elements into account or not. Moreover, note that the `filterUniqReads` argument is set to `TRUE` in order to apply the filtering criteria to unique reads, apart from multi-mapping reads, as proposed by ERVmap authors [@tokuyama2018ervmap]. Conversely, when `filterUniqReads = FALSE`, all the unique reads are considered when computing TE expression without previously filtering them. However, in order for `atena` to be able to distinguish between unique and multi-mapping reads when `filterUniqReads = FALSE`, the NH tag must be present in the BAM file.


```{r, message=FALSE, warning=FALSE}
library(atena)
```


```{r}
bamfiles <- list.files(system.file("extdata", package="atena"),
                       pattern="*.bam", full.names=TRUE)
annot <- ERVmap_ann()
empar <- ERVmapParam(bamfiles, teFeatures = annot, singleEnd = TRUE, 
                     ignoreStrand = TRUE)
empar
```

In the case of paired-end BAM files, two additional arguments can be specified. One is `strandMode`, which defines the behavior of the strand getter when reading the BAM files with `GAlignmentPairs`. The other argument is `fragments`, which sets the read counting approach to be used. The "original" ERVmap algorithm [@tokuyama2018ervmap] is equivalent to setting `fragments = TRUE`, which filters and counts each mate of a paired-end read independently (i.e. two mates mapping to the same element add up a count value of two). On the contrary, when `fragments = FALSE`, if one mate survives the filter criteria but the other mate does not, both mates are discarded and not counted. Moreover, when both mates of a paired-end read pass the filtering, if the two mates map to the same element, they are counted as a single hit (i.e. the two mates add up a count value of one).

## Building a parameter object for Telescope

To use the Telescope method for TE expression quantification, the `TelescopeParam()` function is used to build a parameter object of the class `TelescopeParam`. In this case, annotations developed by Telescope authors including 72,169 ranges, corresponding to 28,513 TEs (i.e. ranges with the same locus ID), are used. The `Telescope_ann()` function retrieves these annotations, and the arguments `type` and `version` can be used to set the type of TE and the human genome version of the annotations, respectively. 

Here, default options are set for running Telescope. However, further [options](https://github.com/mlbendall/telescope) can be passed to the Telescope algorithm through the `opts` argument using a `list` object.

```{r}
bamfiles <- list.files(system.file("extdata", package="atena"),
                       pattern="*.bam", full.names=TRUE)
annot <- Telescope_ann()
tspar <- TelescopeParam(bamfiles, annot)
tspar
```

## Building a parameter object for TEtranscripts

Finally, the third method available is TEtranscripts. First, the `TEtranscriptsParam()` function is called to build a parameter object of the class `TEtranscriptsParam`. The `aggregateby` argument, which should be a character vector of column names in the annotation, determines the columns to be used to perform the aggregation of TE expression quantification. In this case, the default option is used: the names of the `annot` object will be used to aggregate quantifications. For this example, the ERVmap annotations including 3,220 HERVs are used.

```{r}
bamfiles <- list.files(system.file("extdata", package="atena"),
                       pattern="*.bam", full.names=TRUE)
annot <- ERVmap_ann()
ttpar <- TEtranscriptsParam(bamfiles, teFeatures = annot, singleEnd = TRUE, 
                            ignoreStrand=TRUE)

ttpar
```


## Quantify TE expression with `qtex()`

Finally, to quantify TE expression we call the `qtex()` method using either an `ERVmapParam` object or a `TelescopeParam` object as argument, depending on the desired quantification algorithm. Additionally, a `data.frame`, or `DataFrame`, object storing phenotypic data can be passed to the `qtex()` function through the `phenodata` argument to be included in the resulting `SummarizedExperiment` object. The `qtex()` method returns a `SummarizedExperiment` object containing the resulting TE quantification. In the case that phenotypic data was provided, the row names of these phenotypic data are set as column names of the output `SummarizedExperiment` object.

In the current example, the call to quantify TE expression using the ERVmap method would be the following:

```{r, results='hide'}
emq <- qtex(empar)
```
```{r}
emq
colSums(assay(emq))
```

In the case of the Telescope method, the call would be as follows:

```{r, results='hide'}
tsq <- qtex(tspar)
```
```{r}
tsq
colSums(assay(tsq))
```


For the TEtranscripts method, TE expression is quantified by using the following call:

```{r, results='hide'}
ttq <- qtex(ttpar)
```
```{r}
ttq
colSums(assay(ttq))
```


# Session information

```{r session_info, cache=FALSE}
sessionInfo()
```

# References
